Testing Optimal Extraction module

rongmon opened the pull request:
Including an optimal extraction module. I want it tested first before creating a script to do it from command line. 
Also after testing the modules can be merged into other functions, but for testing ease, I kept them stand alone.


# Commands:

from kcwitools import extract_optimal as e
xspec,xspec_std=e.extract_optimal_1D(flux, varcube,hdu_hdr,wave,x1,x2,y1,y2,plot=True)

here flux= KCWI flux cube, varcube= KCWI variance cube, hdu_hdr= header
x1,x2,y1,y2= four corners of a rectangular extraction box.

output = two xspec objects, (1) xspec= optimally extracted spectrum, (2) xspec_std = boxcar extraction.

rongmon committed:
Testing Optimal Extraction module

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+from __future__ import print_function
```
best to use all the future pieces here.
See the other modules.

rongmon replied:
Done!

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+		fluxcube: 3D kcwi fluxcube
+		varcube: 3D KCWI variance cube
+		header:  header information
+		wavegrid:  1D kcwi wavelength grid
```
If this can be trivially derived from the header, then no need to pass it in.

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+		header:  header information
+		wavegrid:  1D kcwi wavelength grid
+		
+		x1,x2,y1,y2: Four corners of an aribaty rectangular pseudo-slit for extraction
```
How about pass this is in a tuple instead, i.e. `corners`?
where 
`x1,x2,y1,y2 = corners`

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+	# CHECK CHECK CHECK
+	#THESE VALUES NEED TO BE DOUBLE CHECKED. WHICH ONE SHOULD I USE?
+	gain=header['GAIN2']
```
As these may be useful elsewhere, I suggest we start a `defs.py`
module and put them in there.

rongmon replied:
Done. Please add more stuff there.

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+
+	# Hardcoding some values
+	sigma=5  # Sigma Clipping Value
```
Put these in the call to the method so that they
become optional.

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+	# Create slices within wavelength ranges that are good for KCWI
+	# From now on only deal with these smaller slices
+	slices=np.where((wavegrid >=3500.) & (wavegrid <= 5500.))
```
No need to use `np.where` and code is cleaner and faster without it, ie.

`slices = (wavegrid >=3500.) & (wavegrid <= 5500.)`

and then

`flux = fluxcube[slices,:,:]`

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+
+	# Plot and show the slit position on the white light image
+	if plot==True:
```
`if plot is True`

`==` is being deprecated in this context.

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+	# Plot and show the slit position on the white light image
+	if plot==True:
+		whitelight=np.sum(flux, axis=0)
```
Shouldn't we use our `whitelight` method instead?

rongmon replied:
Done, and added     header['NAXIS3'] = int(sum(slices)), to reflect the new sliced wavegrid

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+	aperture_mask=np.isin(mask, 1)
+	# This is the masked cube that can be used for extraction
+	cube=flux*mask
+	var_origin[~aperture_mask]=np.inf
```
Scares me a bit to see infinities.  How about 0?

rongmon replied:
I did infinities, because later when we do weight, it is 1/variance. Otherwise weights will be infinite. Should I replace it with a large number?

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+	var_origin[~aperture_mask]=np.inf
+
+	#Get the indices of the aperture spaxels at each wavelength
+	aperture_indices = np.where((aperture_mask[900,:,:] == True))
```
Again, try to avoid `np.where` and use `is True`

rongmon replied:
In this instance, it must be np.where, as this is fed into another function spatial_profile.

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+def extract_optimal_1D(fluxcube, varcube,header,wavegrid,x1,x2,y1,y2,outfile=None,verbose=False,plot=False):
+	"""
+
+	Args:
```
pycharm is telling me that this module is full of tabs.
Replace all of those with spaces.  
tabs are problematic.

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+
+	# Do a boxcar extraction
+	spec=r.extract1d(cube)
```
If we are going to use this method, let's use it for
the boxcar too.  Otherwise, I suggest you just put this
one line of code here.

rongmon replied:
Eliminated this code. Not really needed. (one line code anyhow).

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+	# Do a boxcar extraction
+	spec=r.extract1d(cube)
+	var=var_origin+rdnoise**2
```
Are you sure the `var` cube doesn't already contain the rdnoise?
It ought to.  @astronomeara can tell you.

rongmon replied:
Good question. @astronomeara : please let me know. Depending on the answer, I will modify the code to parse the variance arrays correctly.

astronomeara replied:
I see the following in the headers: 
BIASRN1 =              2.77887 / amp1 RN in e- from bias                        
BIASRN2 =              2.77395 / amp2 RN in e- from bias                        
BIASSUB =                    T / bias subtracted?

astronomeara replied:
And that's in the variance cube header

rongmon replied:
Hmm. @astronomeara : Does that mean that the header input going in now should be the header to the variance cube? I guess it should have all the information right?  
Changing the code to reflect this.  Does that mean my line var =var +rdnoise**2 is still true?

rongmon replied:
I tweaked the code now and now if bias subtraction is done previously, set read noise= 0.
@profxj : please check that is is not bonkers.

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+	var=var_origin+rdnoise**2
+
+
+	cube_bad=np.where(~np.isfinite(cube))
```
eliminate `np.where`

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+	cube[cube_bad]=0
+	cube[cube<0]=0
+
+	varbad=np.where(~np.isfinite(var))
```
same.

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+	varbad=np.where(~np.isfinite(var))
+	var[varbad]=np.inf
+
+	sum_spec=np.sum(np.sum(aperture_mask*cube, axis=2), axis=1)
```
This is the `extract1d` method, right?
I suggest you eliminate it.

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+
+
+	# Now starting Optimal Extraction
+	counter=1.0
```
`counter` should be an `int`

rongmon replied:
Done

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+	#Use the cube of pixel weights, if given
+
+	while counter !=0:
+		if verbose == True:
```
`is True`

profxj commented on kcwitools/extract_optimal.py:
```diff
@@ -0,0 +1,162 @@
+		if verbose == True:
+			print("Wavelength Iteration {}".format(c))
+			print("\n\nCONSTRUCTING SPATIAL PROFILE\n")
+		mcube=r.wavelegth_fitter( cube, var, aperture_mask, wave, fit_order, aperture_indices, plot=False, verbose=False)
```
Is this the method for the spatial profile? 
If so, it needs a better name.

rongmon replied:
Renamed it to spatial_profile

profxj commented on kcwitools/optimal_extract_utility.py:
```diff
@@ -0,0 +1,128 @@
+from __future__ import print_function
```
same as above.

And, add at least a line defining what the module is for.

profxj commented on kcwitools/optimal_extract_utility.py:
```diff
@@ -0,0 +1,128 @@
+
+
+
+# Adopted from some online stuff that uses these for optimal extraction
```
Provide the http link.

profxj commented on kcwitools/optimal_extract_utility.py:
```diff
@@ -0,0 +1,128 @@
+def wavelegth_fitter( fluxcube, varcube, aperture_mask, wavelength, order, aperture_indices, plot=False, verbose=False):
+
+	"""
+	Fit polynomials in the wavelength direction of each pixel in an aperture. 
```
This sounds to me that we are fitting a spectrum to an aperture 
which would be odd.  But I'll read on.

rongmon replied:
Renamed to spatial_profile

profxj commented on kcwitools/optimal_extract_utility.py:
```diff
@@ -0,0 +1,128 @@
+	#Lamdas must be scaled to between -1 and 1 for Chebyshev fitting
+	min_val=wavelength.min()
+	max_val=wavelength.max()
+	scaled_l=np.array([2*(lamda-min_val)/(max_val-min_val) -1 for lamda in wavelength])
```
We have pretty good code for fiitting in PypeIt.
Maybe we should make that a dependency for `kcwitools`
and use them..

rongmon replied:
@profxj : I am happy to do that. Could you help rewrite this routine?

profxj commented on kcwitools/optimal_extract_utility.py:
```diff
@@ -0,0 +1,128 @@
+	max_val=wavelength.max()
+	scaled_l=np.array([2*(lamda-min_val)/(max_val-min_val) -1 for lamda in wavelength])
+
+	for n, (i, j) in enumerate(zip(aperture_indices[0], aperture_indices[1])):
```
Ugly for loop..

rongmon replied:
Please edit the routine

profxj commented on kcwitools/optimal_extract_utility.py:
```diff
@@ -0,0 +1,128 @@
+	if verbose==True:
+		print("Normalising Spatially")
+	#Normalise spatially
+	spatial_norm=np.sum(np.sum(wavelength_fits, axis=2), axis=1)
```
Interesting.  Instead of simply smashing the original data cube along the 
wavelengths, we have fit each spaxel independently and then smashed
those fits.  This may have some benefits, e.g. rejecting CRs and other defects
and smoothing things a bit.  It will, however, probably also fit out emission lines
which could dominate the flux.  What is the typical order for the fit?
And we *must* do rejection when fitting to remove CRs.
Use `robust_polyfit` in PypeIt for that.

rongmon replied:
please edit the routine. I think the cubes I am using already does CR rejection. Moreover, the sigma clipping should take them out regardless after this step?

profxj commented on kcwitools/optimal_extract_utility.py:
```diff
@@ -0,0 +1,128 @@
+
+
+
+def extract1d(datacube,verbose=False):
```
Add doc strings for this and the methods that follow.

rongmon replied:
This is depreciated.

profxj reviewed (changes requested):
Well done @rongmon !

I gave you lots of feedback, in part to guide you towards standards that the `astropy` crowd like,
and in part to improve the algorithm.  I'm happy to do some of the lifting related to these comments.
Just let me know.

rongmon committed:
most changes per first PR

rongmon committed:
small changes

rongmon committed:
Checking bias subtraction

Checking if bias is already subtracted. If yes setting readnoise = 0. Please check that this is correct

rongmon committed:
minor tweaks

rongmon committed:
small_tweaks

profxj commented:
This looks great.  Am approving.

profxj commented:
If @astronomeara is happy, then merge @rongmon

astronomeara commented:
Note that some issues are likely with variance being calculaed from fluxed data (icubes) files.  We should implement a routing for fluxing and used the icubed files insttead.

astronomeara commented:
I'm happy.  Great work, Rongmon!  Merge away
